### PR TITLE
Unregister group path from apiserver when the group no longer exists

### PR DESCRIPTION
**Pull request by tnqn**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

After a CRD or an APIService was deleted, the corresponding group was never unregistered. It caused a stale entry to remain in the root path and could potentially lead to memory leak as the groupDiscoveryHandler was never released and the handledGroups was never cleaned up.

```
# curl 127.0.0.1:8001
{
  "paths": [
    "/.well-known/openid-configuration",
    "/api",
    ...
    "/apis/wardle.example.com",         <- group that has been deleted
    ...
    "/readyz/shutdown",
    "/version"
  ]
}
```

The commit implements the cleanup. It tracks each group's usage and unregister the a group when there is no version for this group.

Incidentally, it fixes a test step in TestSampleAPIServer that it validated APIService deletion using unmatched selector, which selected nothing when deleting and listing APIServices.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug where an API group's path was not unregistered from the API server's root paths when the group was deleted.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -144,8 +144,9 @@ type APIAggregator struct {
 
 	// proxyHandlers are the proxy handlers that are currently registered, keyed by apiservice.name
 	proxyHandlers map[string]*proxyHandler
-	// handledGroups are the groups that already have routes
-	handledGroups sets.String
+	// handledGroups are the groups that already have routes. The key is the name of the group and the value is the
+	// versions for the group.
+	handledGroups map[string]sets.Set[string]
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:149]: call it handledGroupVersions?

**Comment by sttts** [comment]
Beside the naming nit, this looks good.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (tnqn) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: Done, thanks.

**Comment by sftim** [comment]
Changelog suggestion
```markdown
Fixed a bug where an API group's path was not unregistered from the API server's root paths when the group was deleted.
```

**Comment by tnqn** [comment]
@sftim updated, thanks.

**Comment by rtheis** [comment]
Hi folks, I'm not sure if you are aware but Kubernetes CNCF conformance testing with the latest version of sonobuoy against a version 1.29.0-alpha.3 fails on the `support the 1.17 Sample API Server using the current Aggregator` test.  References below have led me here and am wonder if any further actions are required to fix this test failure which seems to be related to this PR.

**References:**
- https://github.com/kubernetes/kubernetes/issues/119582
- https://github.com/kubernetes/test-infra/pull/31123

**Comment by tnqn** [comment]
@rtheis It seems the references explained their failures were because they ran latest test code on master against a cluster deployed using code on a release branch. Is it also your case? If your cluster is 1.29.0-alpha.3, it should include the fix according to https://github.com/kubernetes/kubernetes/blob/8254569d50d6bbe8d60bcdfb16eac337f6e60514/CHANGELOG/CHANGELOG-1.29.md?plain=1#L291
Do you have a sonobuoy result bundle? I could take a look if it's related to the PR.

**Comment by rtheis** [comment]
@tnqn attaching the results bundle.

[202311152139_sonobuoy_63a4c965-7c3e-4284-b70a-1f38e8524393.tar.gz](https://github.com/kubernetes/kubernetes/files/13377110/202311152139_sonobuoy_63a4c965-7c3e-4284-b70a-1f38e8524393.tar.gz)

**Comment by tnqn** [comment]
@rtheis No tests has ran in the bundle.
```
Ran 0 of 7407 Specs in 0.578 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 7407 Skipped
PASS
```

**Comment by rtheis** [comment]
@tnqn sorry, I captured the bundle after the retry attempt which as you can see was not processed properly by sonobuoy.  I'll run again to get the bundle for you.

**Comment by rtheis** [comment]
Take 2 ...

[202311162202_sonobuoy_0841a65b-71cb-4735-96c1-080e6c908000.tar.gz](https://github.com/kubernetes/kubernetes/files/13391692/202311162202_sonobuoy_0841a65b-71cb-4735-96c1-080e6c908000.tar.gz)

**Comment by tnqn** [comment]
@rtheis It seems your kube-apiserver isn't running as Pod so sonobuoy didn't get its log. But I got some image versions from e2e log indicating kubelet and kube-proxy are 1.29.0-alpha.2 while the conformance image is 1.29.0-alpha.3. So it seems you ran into the same issue because mismatched test and code:

- KubeletVersion:v1.29.0-alpha.2+IKS
- KubeProxyVersion:v1.29.0-alpha.2+IKS
- registry.k8s.io/conformance:v1.29.0-alpha.3

**Comment by rtheis** [comment]
@tnqn Thanks for taking a look.  I'll test again with worker nodes updates to v1.29.0-alpha.3.

**Comment by rtheis** [comment]
@tnqn I get the same result with updated worker nodes.  I think we need the tests updated within sonobuoy itself.  I suspect this will happen as part of 1.29 GA work or based on requests such as https://kubernetes.slack.com/archives/C78F00H99/p1700149216028779.

```
Summarizing 1 Failure:
  [FAIL] [sig-api-machinery] Aggregator [It] Should be able to support the 1.17 Sample API Server using the current Aggregator [Conformance] [sig-api-machinery, Conformance]
  test/e2e/apimachinery/aggregator.go:756

Ran 386 of 7407 Specs in 6765.363 seconds
FAIL! -- 385 Passed | 1 Failed | 0 Pending | 7021 Skipped
--- FAIL: TestE2E (6765.65s)
FAIL
```

**Comment by tnqn** [comment]
> @tnqn I get the same result with updated worker nodes.

The updated code is in apiserver. If you only upgraded worker node and still run kube-apiserver v1.29.0-alpha.2 but test with conformance v1.29.0-alpha.3, I think it's expected to fail due to the commit.

**Comment by rtheis** [comment]
@tnqn if you run conformance testing against a v1.29.0-alpha.3 cluster using sonobuoy, I suspect that you will see the failure.

**Comment by tnqn** [comment]
> @tnqn if you run conformance testing against a v1.29.0-alpha.3 cluster using sonobuoy, I suspect that you will see the failure.

I tried v1.29.0-alpha.3 with sonobuoy, it indeed failed. The reason is because sonobuoy serviceaccount doesn't have required permission, also explaining why the test doesn't fail in Kubernetes CI jobs. I created #121985 to track it and proposed two ways to fix it, we could have further discussion there. Thanks for reporting it.

**Pull request merged**